source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
- * PRIORITY
- **********************************************************************************************************************/
-static const size_t s_frame_priority_length = 5;
+    /* Write pad length */
+    if (flags & AWS_H2_FRAME_F_PADDED) {
+        all_wrote &= aws_byte_buf_write_u8(output, pad_length);
+    }
 
-int aws_h2_frame_priority_init(struct aws_h2_frame_priority *frame, struct aws_allocator *allocator) {
-    (void)allocator;
+    /* Increment output->len to jump over the body that we already wrote in */
+    AWS_ASSERT(output->buffer + output->len == body_sub_buf.buffer && "Streamed DATA to wrong position");
+    output->len += body_sub_buf.len;
 
-    AWS_ZERO_STRUCT(*frame);
-    frame->base.type = AWS_H2_FRAME_T_PRIORITY;
+    /* Write padding */
+    if (flags & AWS_H2_FRAME_F_PADDED) {
+        all_wrote &= aws_byte_buf_write_u8_n(output, 0, pad_length);
+    }
 
+    AWS_ASSERT(all_wrote);
     return AWS_OP_SUCCESS;
-}
-void aws_h2_frame_priority_clean_up(struct aws_h2_frame_priority *frame) {
-    AWS_PRECONDITION(frame);
-
-    (void)frame;
-}
 
-int aws_h2_frame_priority_encode(
-    struct aws_h2_frame_priority *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
-
-    (void)encoder;
-
-    const size_t output_init_len = output->len;
-
-    /* Write the header data */
-    if (s_frame_prefix_encode(&frame->base, s_frame_priority_length, 0, output)) {
-        goto write_error;
-    }
-
-    /* Write the priority settings */
-    if (s_frame_priority_settings_encode(&frame->priority, output)) {
-        goto write_error;
-    }
+handle_waiting_for_more_space:
+    ENCODER_LOGF(TRACE, encoder, "Insufficient space to encode DATA for stream %" PRIu32 " right now", stream_id);
+    return AWS_OP_SUCCESS;
 
+handle_nothing_to_send_right_now:
+    ENCODER_LOGF(INFO, encoder, "Stream %" PRIu32 " produced 0 bytes of body data", stream_id);
     return AWS_OP_SUCCESS;
 
-write_error:
-    output->len = output_init_len;
-    return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+error:
+    return AWS_OP_ERR;
 }
 
 /***********************************************************************************************************************
- * RST_STREAM
+ * HEADERS / PUSH_PROMISE
  **********************************************************************************************************************/
-static const size_t s_frame_rst_stream_length = 4;
+DEFINE_FRAME_VTABLE(headers);
+
+/* Represents a HEADERS or PUSH_PROMISE frame (followed by zero or more CONTINUATION frames) */
+struct aws_h2_frame_headers {
+    struct aws_h2_frame base;
+
+    /* Common data */
+    const struct aws_http_headers *headers;
+    uint8_t pad_length; /* Set to 0 to disable AWS_H2_FRAME_F_PADDED */
+
+    /* HEADERS-only data */
+    bool end_stream;   /* AWS_H2_FRAME_F_END_STREAM */
```

Comment:
trivial, maybe take struct packing into account here and group the 1 byte sized members into groups of 2 or 4, or move them to the end

Reply:
I shuffled them a bit,
I didn't want to lose the grouping by category